### PR TITLE
Cleanup + small improvements

### PR DESCRIPTION
**Pull request by mlykotom**
- update dependencies
- refactor deprecated APIs
- add sorting micro benchmarks
- add tags for snippets

### CODE REVIEW

**Review by tikurahul — COMMENTED** [review]
You are changing the `variant` being used to `benchmark` from `release`. 

You will need to update the github workflows so we keep them in sync.

File: BenchmarkSample/benchmark/build.gradle [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
+// [START benchmark_apply_plugin]
 apply plugin: 'androidx.benchmark'
+// [END benchmark_apply_plugin]
 
 android {
     compileSdkVersion 31
 
     defaultConfig {
-        minSdkVersion 14
+        minSdkVersion 21
```
Comment (ChrisCraik) [inline BenchmarkSample/benchmark/build.gradle:11]: why this change?

File: BenchmarkSample/benchmark/build.gradle [hunk 2]
```diff
@@ -34,6 +36,7 @@ android {
     }
 }
 
+// [START benchmark_dependency]
```
Comment (ChrisCraik) [inline BenchmarkSample/benchmark/build.gradle:39]: this seems like a large range, since it covers all of test dependencies - is this to be used in guide docs?

File: BenchmarkSample/benchmark/src/androidTest/AndroidManifest.xml [hunk 3]
```diff
@@ -18,6 +18,8 @@
     xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:tools="http://schemas.android.com/tools"
     package="com.example.benchmark.test">
+
+    <!-- [START benchmark_manifest] -->
```
Comment (ChrisCraik) [inline BenchmarkSample/benchmark/src/androidTest/AndroidManifest.xml:22]: Note that this isn't needed anymore in benchmark 1.1, as we set test variant to release via gradle plugin.

May still want to document this in guidance docs, but only for 1.0 users, and probably doesn't need to be linked to sample - can just remove it here.

File: BenchmarkSample/benchmark/src/androidTest/java/com/example/benchmark/BitmapBenchmark.kt [hunk 4]
```diff
@@ -38,14 +37,14 @@ class BitmapBenchmark {
...
-    private val context = ApplicationProvider.getApplicationContext<Context>()
+    private val context = InstrumentationRegistry.getInstrumentation().targetContext
     private lateinit var bitmap: Bitmap
 
     @Before
     fun setUp() {
-        val inputStream = context.assets.open(JETPACK)
-        bitmap = BitmapFactory.decodeStream(inputStream)
-        inputStream.close()
+        context.assets.open(JETPACK).use {
```
Comment (ChrisCraik) [inline BenchmarkSample/benchmark/src/androidTest/java/com/example/benchmark/BitmapBenchmark.kt:45]: consider just putting this in the bitmap val declaration for simplicity

File: BenchmarkSample/ui/src/main/java/com/example/benchmark/ui/MainActivity.kt [hunk 5]
```diff
@@ -24,35 +23,10 @@ class MainActivity : AppCompatActivity() {
...
-
-private object StringDiffCallback : DiffUtil.ItemCallback<String>() {
-    override fun areItemsTheSame(oldItem: String, newItem: String): Boolean =
-        oldItem == newItem
+// stable random for having the same results
+private val random = Random(0)
 
-    override fun areContentsTheSame(oldItem: String, newItem: String): Boolean =
-        oldItem == newItem
+fun doSomeWork() {
```
Comment (ChrisCraik) [inline BenchmarkSample/ui/src/main/java/com/example/benchmark/ui/MainActivity.kt:29]: this shouldn't be in the activity

File: BenchmarkSample/ui/src/main/java/com/example/benchmark/ui/MainActivity.kt [hunk 6]
```diff
@@ -24,35 +23,10 @@ class MainActivity : AppCompatActivity() {
...
-    override fun onBindViewHolder(holder: Holder, position: Int) {
-        holder.bind(getItem(position))
-    }
-}
-
-private object StringDiffCallback : DiffUtil.ItemCallback<String>() {
-    override fun areItemsTheSame(oldItem: String, newItem: String): Boolean =
-        oldItem == newItem
+// stable random for having the same results
+private val random = Random(0)
```
Comment (ChrisCraik) [inline BenchmarkSample/ui/src/main/java/com/example/benchmark/ui/MainActivity.kt:27]: this randomness isn't stable for two reasons:

1) benchmark will run an arbitrary number of loops. Because this work is inconsistent, the number of loops will affect measured number

2) random isn't reset between benchmarks, so it's easy to have state leaked between them.


randomness in benchmarks should either be:

A) Static, stably seeded data created before the benchmark (each iteration sees same exact input)
B) Random seeded per iteration (functionally same - resulting in the same data each iteration)

File: BenchmarkSample/ui/src/main/java/com/example/benchmark/ui/SampleAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+        return Holder.create(parent)
+    }
+
+    override fun onBindViewHolder(holder: Holder, position: Int) {
+        holder.bind(getItem(position))
+    }
+}
+
+private object StringDiffCallback : DiffUtil.ItemCallback<String>() {
+    override fun areItemsTheSame(oldItem: String, newItem: String): Boolean =
```
Comment (ChrisCraik) [inline BenchmarkSample/ui/src/main/java/com/example/benchmark/ui/SampleAdapter.kt:18]: add comment along the lines of:

in a real diff callback, you'd want to return true if two objects refer to same object, and should e.g. crossfade. In this sample, we don't have changing list item, so simply use equality

File: MacrobenchmarkSample/com.example.macrobenchmark-benchmarkData.json [hunk 8]
```diff
@@ -0,0 +1,40 @@
+{
```
Comment (ChrisCraik) [inline MacrobenchmarkSample/com.example.macrobenchmark-benchmarkData.json:1]: did you mean to check this in? it will get stale very quickly when any benchmarks change

File: MacrobenchmarkSample/gradle.properties [hunk 9]
```diff
@@ -33,3 +33,4 @@ org.gradle.jvmargs=-Xmx2048m -Dfile.encoding=UTF-8
 android.useAndroidX=true
 # Kotlin code style for this project: "official" or "obsolete":
 kotlin.code.style=official
+android.enableAdditionalTestOutput=true
```
Comment (ChrisCraik) [inline MacrobenchmarkSample/gradle.properties:36]: this should be the default for at least arctic fox - can you check if this is needed?

File: MacrobenchmarkSample/macrobenchmark/build.gradle [hunk 10]
```diff
@@ -19,32 +21,39 @@ android {
...
+    // Note that your module name may have different name
+    targetProjectPath = ":app"
+    // Enable the benchmark to run separately from the app process
+    experimentalProperties["android.experimental.self-instrumenting"] = true
 
     buildTypes {
-        // declare a build type (release) to match the target app's build type
-        release {
+        // declare a build type to match the target app's build type
+        benchmark {
```
Comment (ChrisCraik) [inline MacrobenchmarkSample/macrobenchmark/build.gradle:32]: thanks for this - appreciate that it matches our module template now :+1:

File: MacrobenchmarkSample/macrobenchmark/src/main/java/com/example/macrobenchmark/startup/SampleStartupBenchmark.kt [hunk 11]
```diff
@@ -14,37 +14,32 @@
...
     val benchmarkRule = MacrobenchmarkRule()
 
     @Test
-    fun startup() = benchmarkRule.measureStartup(
-        profileCompiled = true,
-        startupMode = startupMode
+    fun startup() = benchmarkRule.measureRepeated(
+        packageName = TARGET_PACKAGE,
+        metrics = listOf(StartupTimingMetric()),
+        iterations = 3,
```
Comment (ChrisCraik) [inline MacrobenchmarkSample/macrobenchmark/src/main/java/com/example/macrobenchmark/startup/SampleStartupBenchmark.kt:39]: increase iterations here and elsewhere - I think 5 is a good minimum that's still fairly responsive when devs run it (10 is what we use within jetpack macrobenchmarks)

File: pull_traces.sh [hunk 12]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+# The following command pulls all files ending in .perfetto-trace from the directory
+# hierarchy starting at the root /storage/emulated/0/Android.
+adb shell find /storage/emulated/0/Android/ -name "*.perfetto-trace" \
```
Comment (ChrisCraik) [inline pull_traces.sh:5]: do you find this script to be useful as is, since benchmark will happily keep creating new trace files, with new names?

was curious about that, when reviewing docs

File: MacrobenchmarkSample/gradle.properties (hunk 9 above)
Comment (mlykotom) [reply MacrobenchmarkSample/gradle.properties:36]: Hmm, it behaves weird, If I run it with `./gradlew :benchmark:cC`
It copies all the .perfetto-trace files (not the benchmarkData.json files), but when it runs `benchmarkReport` task it clears the folder and won't copy anything so there are no outputs.

File: pull_traces.sh (hunk 12 above)
Comment (mlykotom) [reply pull_traces.sh:5]: I thought it may be useful for folks checking the samples and directly pulling the things.
But I think if AS or gradle copies the files properly (the comment above), it may not be needed 🤔

File: MacrobenchmarkSample/gradle.properties (hunk 9 above)
Comment (mlykotom) [reply MacrobenchmarkSample/gradle.properties]: If I run it for macrobenchmark, then when running from AS, it doesn't copy anything (with or without the flag), but running from gradle, it copies properly (without the flag)

File: BenchmarkSample/benchmark/src/androidTest/AndroidManifest.xml (hunk 3 above)
Comment (ChrisCraik) [reply BenchmarkSample/benchmark/src/androidTest/AndroidManifest.xml:18]: It's different between microbench and macrobench. In microbench (this manifest), starting in 1.1:

- You aren't expected to create a benchmark buildType
- Plugin defaults you to the release buildType for tests (which sets debuggable false, and uses release library dependencies (important for avoiding coverage codegen))
- Profileable is only needed for profiling feature

Back in 1.0, I think you would need to force debuggable=false like this, and we hadn't solved the default test type = release, but I think we can just leave this out given how close we are to stable 1.1

I think it's way simpler to avoid differentiating 1.0, given how much it'll complicate things for such a short time

File: BenchmarkSample/ui/src/main/java/com/example/benchmark/ui/MainActivity.kt [hunk 13]
```diff
@@ -1,17 +1,16 @@
...
 import androidx.appcompat.app.AppCompatActivity
-import androidx.recyclerview.widget.DiffUtil
-import androidx.recyclerview.widget.ListAdapter
+import androidx.lifecycle.ViewModel
 import androidx.recyclerview.widget.RecyclerView
 
 class MainActivity : AppCompatActivity() {
+
+    // It's used in benchmarks
+    @Suppress("MemberVisibilityCanBePrivate")
```
Comment (ChrisCraik) [inline BenchmarkSample/ui/src/main/java/com/example/benchmark/ui/MainActivity.kt:11]: @TestOnly?

File: MacrobenchmarkSample/com.example.macrobenchmark-benchmarkData.json (hunk 8 above)
Comment (ChrisCraik) [reply MacrobenchmarkSample/com.example.macrobenchmark-benchmarkData.json:1]: ACK

**Review by ChrisCraik — APPROVED** [review]
Generally LGTM, you can make the call on the unresolved bits.

Have you tested with github action, that everything continues to build and run? (I don't remember if that's postsubmit only)

**Review by tikurahul — APPROVED** [review]

File: BenchmarkSample/ui/src/main/java/com/example/benchmark/ui/MainActivity.kt (hunk 13 above)
Comment (mlykotom) [reply BenchmarkSample/ui/src/main/java/com/example/benchmark/ui/MainActivity.kt:11]: hm, is it TestOnly? It's called from the class itself as well.

**Comment by mlykotom** [comment]
I ran the FTL manually and works ok
